lib/push.py

```diff
@@ -81,24 +89,75 @@ def manifest_new(class_):
                "weirdal": "yankovic" }
 
    def cleanup(self):
+      # Delete the config, manifest, and tarballs if the cache is disabled.
+      conf_suffix = str(self.image) + ".config.json"
+      conf_path = ch.storage.upload_cache // conf_suffix
+      man_suffix = str(self.image) + ".manifest.json"
+      man_path = ch.storage.upload_cache // man_suffix
+      if (not conf_path.exists() or not man_path.exists()):
+         return
       ch.INFO("cleaning up")
-      # Delete the tarballs since we can't yet cache them.
+      ch.VERBOSE("deleting config: %s" % conf_path)
+      conf_path.unlink()
+      ch.VERBOSE("deleting manifest: %s" % man_path)
+      man_path.unlink()
       for (_, tar_c) in self.layers:
          ch.VERBOSE("deleting tarball: %s" % tar_c)
          tar_c.unlink_()
 
-   def prepare(self):
-      """Prepare self.image for pushing to self.dst_ref. Return tuple: (list
-         of gzipped layer tarball paths, config as a sequence of bytes,
-         manifest as a sequence of bytes).
+   def rename_conf_and_man(self, git_hash):
+      conf_suffix = str(self.image) + ".config.json"
+      man_suffix = str(self.image) + ".manifest.json"
+      return (conf_suffix, man_suffix)
+
+   def rename_tar(self, git_hash, path_c):
+      return path_c
+
+   def prepare_upload(self, config, manifest, tars_c):
+      """Store a list of gzipped layer tarball paths, config as a sequence of
+       bytes, and manifest as a sequence of bytes."""
+      config_bytes = json.dumps(config, indent=2).encode("UTF-8")
+      config_hash = ch.bytes_hash(config_bytes)
+      manifest["config"]["size"] = len(config_bytes)
+      manifest["config"]["digest"] = "sha256:" + config_hash
+      ch.DEBUG("config: %s\n%s" % (config_hash, config_bytes.decode("UTF-8")))
+      manifest_bytes = json.dumps(manifest, indent=2).encode("UTF-8")
+      ch.DEBUG("manifest:\n%s" % manifest_bytes.decode("UTF-8"))
```

Comment:
With the other suggested changes, we shouldn't need arguments passed to `prepare_upload()`. Here we just real the files from the `self.path_config`, `self.path_manfiest` etc.